### PR TITLE
Fix hard-coded `double` precision in test_functions to default dtype

### PR DESCRIPTION
**Pull request by AVHopp**
## Motivation

This PR replaces the hard-coded double precision that was used in the initialization of `test_functions/base.py` to use `torch.get_default_dtype()` instead.

See #2596 for more details.

### Have you read the [Contributing Guidelines on pull requests](https://github.com/pytorch/botorch/blob/main/CONTRIBUTING.md#pull-requests)?

Yes, I have read it.

## Test Plan

I ran code formatting via `ufmt` and checked the code via `pytest -ra`. All tests related to the changes here were adjusted in the second commit of the branch.
Locally, two tests failed for me, but it seems to me that these are not related to the fix implemented here. If it turns out they are, I'd be more than happy to further adjust.

## Related PRs

None, but #2596 is related.

### CODE REVIEW

**Comment by AVHopp** [comment]
I am aware that I still need to sign the CLA and will do so soon.

**Comment by AVHopp** [comment]
I have signed the CLA, but it seems like the test was not retriggered, does this just take some more time or did something go wrong with me signing it?

**Comment by Balandat** [comment]
> I have signed the CLA, but it seems like the test was not retriggered, does this just take some more time or did something go wrong with me signing it?

No need to re-test. I can now import the PR and land it even if GitHub says that check failed.

**Comment by AVHopp** [comment]
@Balandat @esantorella Pinging both of you since we already started discussions on #2596 .

**TL;DR:** Changed to use `dtype` in `__init__`. Looking forward to feedback, how to deal with the branch history?

As proposed there, I have re-implemented this now such that there is an explicit `dtype` argument in all of the test functions. I just added this as an argument to every explicit `__init__` function, always setting it to `torch.double`. Not sure if there is a more elegant way, if so feel free to point me to it :)

For better reviewing, each commit just touches a single file - however, I'd be more than happy to rebase/squash the commits at a later point, but since I could not find any guidelines regarding whether or not this is done here, I thought I'd better just have a clean but "true" history for now 😄 

I tested the code again locally using `pytest` and also in my own code where I observed this problem and everything seems to work.

Looking forward to your feedback/comments :)

File: botorch/test_functions/base.py [hunk 1]
```diff
@@ -180,8 +185,8 @@ def __init__(
                 f"If specified as a list, length of noise_std ({len(noise_std)}) "
                 f"must match the number of objectives ({len(self._ref_point)})"
             )
-        super().__init__(noise_std=noise_std, negate=negate)
-        ref_point = torch.tensor(self._ref_point, dtype=torch.get_default_dtype())
+        super().__init__(noise_std=noise_std, negate=negate, dtype=dtype)
+        ref_point = torch.tensor(self._ref_point, dtype=self.dtype)
```
Comment (AVHopp) [inline botorch/test_functions/base.py:189]: Note: I spotted another use of `torch.get_default_dtype` while implementing my changes - I guess that this should be changed, only question is to what. For now, I used `self.dtype`, but I guess it should rather be `X.dtype`?

File: botorch/test_functions/base.py [hunk 2]
```diff
@@ -37,17 +38,20 @@ def __init__(
                 provided, specifies separate noise standard deviations for each
                 objective in a multiobjective problem.
             negate: If True, negate the function.
+            dtype: The dtype that is used for the bounds of the function.
         """
         super().__init__()
         self.noise_std = noise_std
         self.negate = negate
+        self.dtype = dtype
```
Comment (Balandat) [inline botorch/test_functions/base.py:46]: why assign this? Seems like the only place we'd need this is to register the `bounds` buffer?

File: botorch/test_functions/base.py (hunk 1 above)
Comment (Balandat) [reply botorch/test_functions/base.py:189]: At this point during init we don't know what `X` will be - so I think the only reasonable thing to do here would be to be consistent with how we handle the `bounds`.

This raises the question what to do when getting an `X` with a different `dtype` during `forward` or `evaluate_true`. But I think we should be able to just rely on type promotion there. 


```suggestion
        ref_point = torch.tensor(self._ref_point, dtype=dtype)
```

File: botorch/test_functions/multi_fidelity.py [hunk 3]
```diff
@@ -74,13 +74,18 @@ class AugmentedHartmann(SyntheticTestFunction):
...
+    def __init__(
+        self,
+        noise_std: float | None = None,
+        negate: bool = False,
+        dtype: torch.dtype = torch.double,
+    ) -> None:
         r"""
         Args:
             noise_std: Standard deviation of the observation noise.
             negate: If True, negate the function.
```
Comment (Balandat) [inline botorch/test_functions/multi_fidelity.py:86]: missing dtype arg in docstring

File: botorch/test_functions/multi_fidelity.py [hunk 4]
```diff
@@ -126,7 +131,11 @@ class AugmentedRosenbrock(SyntheticTestFunction):
     _optimal_value = 0.0
 
     def __init__(
-        self, dim=3, noise_std: float | None = None, negate: bool = False
+        self,
+        dim=3,
+        noise_std: float | None = None,
+        negate: bool = False,
+        dtype: torch.dtype = torch.double,
```
Comment (Balandat) [inline botorch/test_functions/multi_fidelity.py:138]: missing dtype arg in docstring below

File: botorch/test_functions/synthetic.py [hunk 5]
```diff
@@ -261,6 +264,7 @@ def __init__(
         noise_std: float | None = None,
         negate: bool = False,
         bounds: list[tuple[float, float]] | None = None,
+        dtype: torch.dtype = torch.double,
```
Comment (Balandat) [inline botorch/test_functions/synthetic.py:267]: missing dtype arg in docstring below

**Comment by Balandat** [comment]
> For better reviewing, each commit just touches a single file - however, I'd be more than happy to rebase/squash the commits at a later point, but since I could not find any guidelines regarding whether or not this is done here, I thought I'd better just have a clean but "true" history for now 😄

I think it's fine to have these as one commit. But it doesn't really matter for the branch history here; the way we merge things into the main branch in botorch is that our github bot will auto-squash every change from this PR into a single commit.

File: botorch/test_functions/base.py (hunk 2 above)
Comment (AVHopp) [reply botorch/test_functions/base.py:46]: You are right, there is no reason for this. Will remove, commit and resolve this comment then.

Comment (AVHopp) [reply botorch/test_functions/base.py]: Removed the assignment, hence resolving this comment.

File: botorch/test_functions/multi_fidelity.py (hunk 3 above)
Comment (AVHopp) [reply botorch/test_functions/multi_fidelity.py:86]: Added the argument to the docstring, hence resolving this conversation.

File: botorch/test_functions/multi_fidelity.py (hunk 4 above)
Comment (AVHopp) [reply botorch/test_functions/multi_fidelity.py:139]: Adding argument to docstring, hence resolving this conversation.

File: botorch/test_functions/synthetic.py (hunk 5 above)
Comment (AVHopp) [reply botorch/test_functions/synthetic.py:267]: Added the docstring, hence resolving this conversation.

**Review by Balandat — APPROVED** [review]

**Comment by Balandat** [comment]
Thanks!
